batchflow/models/torch/base.py

```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
                 p.grad.add_(previous_grad)
 
 
-    def predict(self, inputs, targets=None, outputs=None, lock=True):
+    def predict(self, inputs, targets=None, outputs=None, lock=True, microbatch=False):
         """ Get predictions on the data provided.
 
         Parameters
         ----------
-        !!
-        targets : ndarray, optional
-            Targets to calculate loss.
-        fetches : tuple, list
-            Sequence of tensors to fetch from the model.
+        inputs : np.ndarray or sequence of them
+            Model inputs. Passed directly to model.
+        targets : np.ndarray or sequence of them
+            Optional model targets to calculate loss with. Passed directly to model.
+        outputs : str or sequence of them
+            Desired outputs of the method.
+            Each string defines a tensor to get and should be one of:
+                - pre-defined tensors, which are `predictions`, `loss`, and `predictions_{i}` for multi-output models.
+                - values described in the `outputs` key in the config
+                - layer id, which describes how to access the layer through a series of `getattr` and `getitem` calls.
+                Allows to get intermediate activations of a neural network.
         lock : bool
             If True, then model and loss computation operations are locked, thus allowing for multithreading.
-        kwargs : dict
-            Additional named arguments directly passed to `feed_dict`.
+        microbatch : int, bool or None
+            If int, then size of chunks to split every batch into. Allows to process given data sequentially.
+            If True, then value from config is used (default value is not to use microbatching).
+            If False or None, then microbatching is not used.
 
         Returns
         -------
-        Calculated values of tensors in `fetches` in the same order.
+        Calculated values of tensors in `outputs` in the same order.
 
         Examples
         --------
-        .. code-block:: python
+        Layer ids allow to get intermediate activations. If the model has `batchflow_model.model.head[0]` layer,
+        you can access it with::
+
+        >>> batchflow_model.predict(inputs=B.images, outputs='model.head[0]')
 
-            model.predict(B('images'), targets=B('labels'), fetches='loss')
+        String keys for `getitem` calls are also allowed::
+
+        >>> batchflow_model.predict(inputs=B.images, outputs='model.body.encoder["block-0"]')
         """
         # Acquire lock; release in any case
         try:
             if lock:
                 self.model_lock.acquire()
+            self.last_predict_info = {}
+
+            # Parse inputs and targets: always a list
+            inputs = list(inputs) if isinstance(inputs, (tuple, list)) else [inputs]
```

Comment:
to_list?

https://github.com/analysiscenter/batchflow/blob/7d201bfdf220496e966e9db5eb27a57212638994/batchflow/utils.py#L19

Reply:
I really don't like the implementation of it, as it breaks in some cases (although not really relevant here until we try to pass anything but `np.ndarray`s)

Also, it would change one clear pythonic line to just another line of code with a function call, which may do unexpected things (and in the case it indeed would do that)